### ISSUE TITLE
Poll TradeOffer status automatically

### ISSUE THREAD

**Issue by BlueRaja** [Jessecar96/SteamBot#914]
So, I'm finally _finally_ getting around to looking at our code for Trade Offers.  As far as I can tell from the code, this is the current typical use-case:

1. When we get a `SteamBot.SteamNotifications.NotificationCallback` from SteamKit, we call `TradeOfferManager.GetOffers()`.
2. This makes a web request to the Steam [GetTradeOffers](https://developer.valvesoftware.com/wiki/Steam_Web_API/IEconService#GetTradeOffers_.28v1.29) API, to obtain information about all current _(and past, up to 1000)_ trade offers.
3. Any active trade offers that haven't been seen before _(since the bot started up)_ call `UserHandler.OnNewTradeOffer(TradeOffer)`
4. The user is now expected to poll `TradeOfferManager.GetOffer()` with the `TradeOffer.TradeOfferId` for that trade until it completes.

Additionally, there is `UserHandler.OnTradeAwaitingConfirmation(tradeOfferId)` which gets called in the specific case that a normal trade turns into a trade offer _(shouldn't that be 100% of trades now, since either you have mobile confirmations or you have an escrow period?)_.  I believe `UserHandler.OnNewTradeOffer(TradeOffer)` will still also be called in this case, usually.

## My Suggestion  ##
_(edited to include below suggestion)_

* ~~Remove `UserHandler.OnTradeAwaitingConfirmation()`, since it is redundant with `UserHandler.OnNewTradeOffer()`~~ _(nevermind, bots need to know the `tradeofferid` of finished trades)_
* Remove `UserHandler.OnTradeSuccess()` since it can never be called now (...right?)
* Replace `UserHandler.OnNewTradeOffer()` with `UserHandler.OnTradeOfferUpdated()`
* Edit `TradeOfferManager.GetOffers()` _(plural)_ to call the new method on all trade offer updates.
* Have the bot poll `TradeOfferManager.GetOffers()` automatically.  We can add the poll-time as a config parameter.  There is no need to call `TradeOfferManager.GetOffer()` _(singular)_ on each trade anymore, because that information is all available in `TradeOfferManager.GetOffers()`

I don't mind doing the work for this, but I wanted to hear your guys' opinions on the proposal first.

**Comment by scholtzm** [comment]
This is pretty much how [node-steam-tradeoffer-manager](https://github.com/DoctorMcKay/node-steam-tradeoffer-manager) works and it's pretty excellent.

**Comment by BlueRaja** [comment]
Another option would be to remove `UserHandler.OnNewTradeOffer()` and replace it with `UserHandler.OnTradeOfferStatusUpdated()` _(without adding any other new methods)_.

This may actually be better, because in my original suggestion, `UserHandler.OnTradeOfferSuccess()` would get called multiple times if the bot restarts.  In this case, it'd be much more clear that the developer needs to keep track of which trade offers have already completed themselves.

**Comment by waylaidwanderer** [comment]
This is the loop I wrote for checking trade offer statuses: https://github.com/waylaidwanderer/SteamTradeOffersBot/blob/SteamTradeOffersBot/SteamAPI/TradeOffers/TradeOffers.cs#L371

**Comment by ThereWasADream** [comment]
Unless I'm doing something wrong, OnTradeAwaitingConfirmation is never called anymore. I've had my bot log this automatically for ages and it's still never shown up in my log.

**Comment by BlueRaja** [comment]
oh wow I've been putting this off for a month.  I better get started on this for real-real

**Comment by ThereWasADream** [comment]
@BlueRaja Thanks for your efforts! We all have busy lives.  :+1:

**Comment by waylaidwanderer** [comment]
In the meantime, [SteamTradeOffersBot](https://github.com/waylaidwanderer/SteamTradeOffersBot/blob/SteamTradeOffersBot) works great if you need to handle callbacks for trade offer status changes.

**Comment by BlueRaja** [comment]
Is that something we can merge back into SteamBot?  Or are the changes too large?

**Comment by waylaidwanderer** [comment]
It's using my own trade offers class and the design is very different. An example is you have to catch exceptions when accepting trades, since instead of returning false you must actually handle the Steam error given (see [here](https://github.com/waylaidwanderer/SteamTradeOffersBot/blob/SteamTradeOffersBot/SteamBot/TradeOfferUserHandler.cs#L24-L39)). I've also added a lot of my own stuff to SteamTradeOffersBot (such as a nice interface for generic inventories and some quality-of-life changes).
In short it's a lot of code-breaking and design-breaking changes that I felt wouldn't fit into SteamBot, which is why it works better as a fork. I wouldn't mind merging all that stuff into SteamBot, but it's definitely not backwards-compatible.

**Comment by BlueRaja** [comment]
@waylaidwanderer The changes I'm suggesting here would also break backwards compatibility, though I suppose it's a matter of how much.

This repo is also long overdue for an Inventory rewrite, which would also break backwards compatibility.  I made a list of goals a long time ago [here](https://github.com/Jessecar96/SteamBot/issues/642) - if we could fulfill those goals, I think it would be totally worth making a major breaking-change _(maybe calling it "SteamBot 2.0?)_.  Does your new inventory system meet those goals?

**Comment by waylaidwanderer** [comment]
Take a look at my example for working with generic inventories: https://github.com/waylaidwanderer/SteamTradeOffersBot/blob/SteamTradeOffersBot/SteamBot/TradeOfferUserHandler.cs#L137-L165

The class itself is here: https://github.com/waylaidwanderer/SteamTradeOffersBot/blob/SteamTradeOffersBot/SteamAPI/GenericInventory.cs

It does seem to meet quite a few of your goals though, skimming over it.

**Comment by BlueRaja** [comment]
I have some large security concerns about the multithreading code, but other than that it looks good.  You seem to be the most active SteamBot developer currently, so I think merging this in and calling it "SteamBot 2.0" is our best bet of moving SteamBot forward.

However of course I can't make that call myself.  What does everything else think?

Also @waylaidwanderer I noticed the repo is 50-some commits behind.  How horrible are the merge conflicts if you merge in from develop, so SteamTradeOffersBot has all the latest fixes?

**Comment by waylaidwanderer** [comment]
I don't think merging would work, especially since I stripped out a lot of the SteamBot code pertaining to live trades.

I could probably make another fork and put in all my SteamTradeOffersBot changes on top of SteamBot though.

**Comment by BlueRaja** [comment]
You could also cherry-pick the commits which are actually relevant to your branch.  Anything not related to trade offers, I assume.

**Comment by waylaidwanderer** [comment]
The issue is that I've removed all live trading stuff from my fork, so you can only trade via trade offers.

**Comment by Bottswana** [comment]
If we can come up with some tasks for this 2.0 release that need
completing, I can put aside some time to assist.

-J

**Comment by BlueRaja** [comment]
> The issue is that I've removed all live trading stuff from my fork

Do you mean you've removed things like `OnTradeSuccess()`, which no longer works anyways?  Or you've really removed **all** live-trading stuff, like `OnTradeItemAdded()`?  If it's the latter, I don't think it will be acceptable for SteamBot 2.0 then :\

**Comment by scholtzm** [comment]
It doesn't contain anything related to "live" trades as far as I can tell.

**Comment by waylaidwanderer** [comment]
@BlueRaja the latter. That's why I said "I could probably make another fork and put in all my SteamTradeOffersBot changes on top of SteamBot though." Minus the removal of live trading, of course.

**Comment by BlueRaja** [comment]
Ok, please do that ^_^ It's in everyone's best interest if we can avoid fragmenting the community, the more people using/fixing bugs/adding features to the same repo, the better.

**Comment by waylaidwanderer** [comment]
I'm actually quite busy with some freelance projects I'm working on, so I can't promise it'll be done anytime soon though.

**Comment by BlueRaja** [comment]
Hrm, I was planning on setting aside this entire weekend to work on this.

Would you have any time to work with me this weekend to get the transfer started?

**Comment by waylaidwanderer** [comment]
I doubt it. My free time will be sparse until after April.

**Comment by BlueRaja** [comment]
I may just make the changes above separately then :( Unless anyone else is familiar with @waylaidwanderer's bot and would like to help?

**Comment by BlueRaja** [comment]
Completed with #934
